storage: don't perform one-phase commit transactions after restarts

nvb opened the pull request:
Fixes #40466.

This commit adjusts the Replica-side logic around executing one-phase commit transactions. It prevents batches that could be executed on the 1PC fast-path from doing so if they contain transactions in any epoch other than their first epoch. We saw in #40466 that this could lead to intent abandonment if the txn had written intents in earlier epochs.

I believe that this is a new issue that was introduced when we moved from using the presence of a BeginTxn and EndTxn in the same batch to detect 1PC transactions to using the sequence numbers of writes and the presence of an EndTxn in a batch to detect 1PC transactions. That change was necessary for parallel commits. Before that change, I think logic in DistSender was preventing us from hitting this case because it would always split EndTxn requests from the rest of its batch when a txn had been restarted: https://github.com/cockroachdb/cockroach/blob/18bdfe1a691fa6d785d510e86d27cecdac9c436e/pkg/kv/dist_sender.go#L692-L697

nvb committed:
storage: remove StoreTestingKnobs.DisableOptional1PC

This was no longer in use, as of 374413f3aec9150dc57b897a366010dcc99184f6.

Release note: None

nvb committed:
storage: don't perform one-phase commit transactions after restarts

Fixes #40466.

This commit adjusts the Replica-side logic around executing one-phase commit
transactions. It prevents batches that could be executed on the 1PC fast-path
from doing so if they contain transactions in any epoch other than their first
epoch. We saw in #40466 that this could lead to intent abandonment if the txn
had written intents in earlier epochs.

I believe that this is a new issue that was introduced when we moved from using
the presence of a BeginTxn and EndTxn in the same batch to detect 1PC transactions
to using the sequence numbers of writes and the presence of an EndTxn in a batch
to detect 1PC transactions. That change was necessary for parallel commits. Before
that change, I think logic in DistSender was preventing us from hitting this case
because it would always split EndTxn requests from the rest of its batch when a
txn had been restarted:
 https://github.com/cockroachdb/cockroach/blob/18bdfe1a691fa6d785d510e86d27cecdac9c436e/pkg/kv/dist_sender.go#L692-L697

Release note: None

nvb committed:
storage: set request sequence numbers in TestIsOnePhaseCommit

This change correctly sets sequence numbers in TestIsOnePhaseCommit
so that we can test the detection of one-phase commit transactions
whose batches do not contain BeginTransaction requests.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/40518)
<!-- Reviewable:end -->

nvb committed:
storage: don't check txn deadline twice in isOnePhaseCommit

As of #35284, this was already checked by IsEndTransactionTriggeringRetryError.

Release note: None

nvb committed:
storage: export CanForwardCommitTimestampWithoutRefresh

We can then use this to centralize its logic and avoid leaking it
around 1PC transaction handling.

Release note: None

tbg reviewed (approved):
> Reviewed 3 of 3 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40518)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
bors r+
